qiskit/synthesis/su4/siswap_decompose.py

```diff
@@ -0,0 +1,286 @@
+            follow = QuantumCircuit(2)
+
+            # x -> x + pi/8
+            # y -> y + pi/8
+            follow = follow.compose(SiSwapGate(), [0, 1])
+
+            eigenphase_crossing = False
+            if x > np.pi / 8:  # green region
+                # y -> y + pi/8
+                # z -> z + pi/8
+                follow = follow.compose(YGate().power(1 / 2), [0], front=True)
+                follow = follow.compose(YGate().power(1 / 2), [1], front=True)
+                follow = follow.compose(YGate().power(-1 / 2), [0])
+                follow = follow.compose(YGate().power(-1 / 2), [1])
+                if y + z < np.pi / 4:  # eigenphase crossing condition: a_2 - pi/4 < a_3 + pi/4
+                    eigenphase_crossing = True
+                # corresponding red coordinates
+                y -= np.pi / 8
+                z -= np.pi / 8
+            else:  # purple region
+                # x -> x - pi/8
+                # z -> z + pi/8
+                follow = follow.compose(XGate().power(1 / 2), [0], front=True)
+                follow = follow.compose(XGate().power(1 / 2), [1], front=True)
+                follow = follow.compose(XGate().power(-1 / 2), [0])
+                follow = follow.compose(XGate().power(-1 / 2), [1])
+                follow = follow.compose(ZGate(), [0], front=True)
+                follow = follow.compose(ZGate(), [0])
+                if y + z < np.pi / 8:  # eigenphase crossing condition: a_2 - pi/4 < a_3
+                    eigenphase_crossing = True
+                # corresponding red coordinates
+                x += np.pi / 8
+                z -= np.pi / 8
+
+            if eigenphase_crossing:
+                y, z = -z, -y
+
+            # final 3xSiSwap circuit: red --> fix crossing --> green or purple
+            red_decomp = _can_circuit(x, y, z)
+            circuit = circuit.compose(red_decomp, [0, 1])
+            if eigenphase_crossing:
+                # y, z -> -z, -y
+                circuit = circuit.compose(XGate().power(1 / 2), [0], front=True)
+                circuit = circuit.compose(XGate().power(1 / 2), [1], front=True)
+                circuit = circuit.compose(XGate().power(-1 / 2), [0])
+                circuit = circuit.compose(XGate().power(-1 / 2), [1])
+                circuit = circuit.compose(XGate(), [0], front=True)
+                circuit = circuit.compose(XGate(), [0])
+            circuit = circuit.compose(follow, [0, 1])
+
+    return circuit
+
+
+def _interleaving_single_qubit_gates(x, y, z):
+    """
+    Find the single-qubit gates given the interaction coefficients
+    (x, y, z) ∈ W′ when sandwiched by two SiSwap gates.
+    Return the SiSwap sandwich.
+    """
+    C = np.sin(x + y - z) * np.sin(x - y + z) * np.sin(-x - y - z) * np.sin(-x + y + z)
+    if abs(C) < _EPS:
+        C = 0.0
+
+    α = np.arccos(np.cos(2 * x) - np.cos(2 * y) + np.cos(2 * z) + 2 * np.sqrt(C))
```

Comment:
Wherever you're dealing with scalars and not things that might be Numpy arrays to be vectorised over, it's better to use the functions from `math` or `cmath` as appropriate rather than `numpy`, because `numpy` ones have a really very sizeable dispatch overheard when used on scalars.

Reply:
I knew it was there, but I'm surprised how large it is. I just did a test with `npsin = np.sin` and `mathsin = math.sin`. timeit showed numpy more than 10x slower.